### ISSUE TITLE
Use Catch for unit tests

### ISSUE THREAD

**Issue by nlohmann** [nlohmann/json#3232]
The latest Doctest release added a regression with MSVC 2015 (see #3228 and https://github.com/doctest/doctest/issues/573). Furthermore, Doctest is looking for maintainers (https://github.com/doctest/doctest/issues/554) and the future of the project is unclear. Meanwhile, Catch seems to be actively developed.

**Comment by philbucher** [comment]
Hi @nlohmann,
did you measure the compilation and runtimes of the two libraries?
I switched from Catch to Doctest for one of my projects some time ago because the compilation times were comparatively large with Catch.
However this was before Catch decided to drop the header-only approach, so not sure how the situation looks currently

just curious, as I am somewhat involved with doctest now :)

**Comment by nlohmann** [comment]
This was the same reason for me to switch to Doctest. I don't really want to switch back to Catch, but I was afraid that Doctest somehow stalled as they were looking for maintainers (which seems to have been solved now). I did not compare the compilation times yet - the Catch branch #3236 does not work right now, because Catch triggers several compiler warnings. If the newest Doctest release 2.4.8 works, I will close this issue and hope to not touch the test suite again for a long time :)
